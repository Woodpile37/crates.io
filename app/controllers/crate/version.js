import Controller from '@ember/controller';
import { action } from '@ember/object';
import { inject as service } from '@ember/service';
import { tracked } from '@glimmer/tracking';

import { task } from 'ember-concurrency';
import { alias } from 'macro-decorators';

export default class CrateVersionController extends Controller {
  @service session;

  get downloadsContext() {
    return this.requestedVersion ? this.currentVersion : this.crate;
  }

  @tracked stackedGraph = true;

  @action setStackedGraph() {
    this.stackedGraph = true;
  }

  @action setUnstackedGraph() {
    this.stackedGraph = false;
  }

  @alias('downloadsContext.version_downloads.content') downloads;
  @alias('model.crate') crate;
  @alias('model.requestedVersion') requestedVersion;
  @alias('model.version') currentVersion;

  get isOwner() {
    return this.crate.owner_user.findBy('id', this.session.currentUser?.id);
  }

  @alias('loadReadmeTask.last.value') readme;

  loadReadmeTask = task(async () => {
    let version = this.currentVersion;

    let readme = version.loadReadmeTask.lastSuccessful
      ? version.loadReadmeTask.lastSuccessful.value
      : await version.loadReadmeTask.perform();

    if (typeof document !== 'undefined') {
      setTimeout(() => {
        let e = document.createEvent('CustomEvent');
        e.initCustomEvent('hashchange', true, true);
        window.dispatchEvent(e);
      });
    }

    return readme;
  });
}
