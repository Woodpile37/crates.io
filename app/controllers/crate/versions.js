import Controller from '@ember/controller';
import { tracked } from '@glimmer/tracking';

export default class SearchController extends Controller {
  queryParams = ['sort'];

  @tracked sort;

  get currentSortBy() {
    return this.sort === 'semver' ? 'SemVer' : 'Date';
  }

  get sortedVersions() {
    let versions = this.model.versions.toArray();

    return this.sort === 'semver'
      ? versions.sort(compareBySemver)
      : versions.sort((a, b) => b.created_at - a.created_at);
  }
}

function compareBySemver(a, b) {
  let aSemver = a.semver;
  let bSemver = b.semver;

  if (aSemver === bSemver) {
    return b.created_at - a.created_at;
  } else if (aSemver === null) {
    return 1;
  } else if (bSemver === null) {
    return -1;
  } else {
    return bSemver.compare(aSemver);
  }
}
