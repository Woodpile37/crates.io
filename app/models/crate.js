import Model, { attr, hasMany } from '@ember-data/model';
import { waitForPromise } from '@ember/test-waiters';

import { apiAction } from '@mainmatter/ember-api-actions';

export default class Crate extends Model {
  @attr name;
  @attr downloads;
  @attr recent_downloads;
  @attr('date') created_at;
  @attr('date') updated_at;
  @attr max_version;
  @attr max_stable_version;
  @attr newest_version;

  @attr description;
  @attr homepage;
  @attr documentation;
  @attr repository;

  @hasMany('version', { async: true, inverse: 'crate' }) versions;
  @hasMany('team', { async: true, inverse: null }) owner_team;
  @hasMany('user', { async: true, inverse: null }) owner_user;
  @hasMany('version-download', { async: true, inverse: null }) version_downloads;
  @hasMany('keyword', { async: true, inverse: null }) keywords;
  @hasMany('category', { async: true, inverse: null }) categories;
  @hasMany('dependency', { async: true, inverse: null }) reverse_dependencies;

  /**
   * This is the default version that will be shown when visiting the crate
   * details page. Note that this can be `undefined` if all versions of the crate
   * have been yanked.
   * @return {string}
   */
  get defaultVersion() {
    if (this.max_stable_version) {
      return this.max_stable_version;
    }
    if (this.max_version && this.max_version !== '0.0.0') {
      return this.max_version;
    }
  }

  get owners() {
    let teams = this.owner_team.toArray() ?? [];
    let users = this.owner_user.toArray() ?? [];
    return [...teams, ...users];
  }

  async follow() {
    return await waitForPromise(apiAction(this, { method: 'PUT', path: 'follow' }));
  }

  async unfollow() {
    return await waitForPromise(apiAction(this, { method: 'DELETE', path: 'follow' }));
  }

  async inviteOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'PUT', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  async removeOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'DELETE', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }
}
