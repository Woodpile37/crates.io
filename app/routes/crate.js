import { NotFoundError } from '@ember-data/adapter/error';
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';

export default class CrateRoute extends Route {
  @service headData;
  @service router;
  @service store;

  async model(params, transition) {
    let crateName = params.crate_id;

    try {
      return await this.store.findRecord('crate', crateName);
    } catch (error) {
      if (error instanceof NotFoundError) {
        let title = `${crateName}: Crate not found`;
        this.router.replaceWith('catch-all', { transition, error, title });
      } else {
        let title = `${crateName}: Failed to load crate data`;
        this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
      }
    }
  }

  setupController(controller, model) {
    super.setupController(...arguments);
    this.headData.crate = model;
  }

  resetController() {
    super.resetController(...arguments);
    this.headData.crate = null;
  }
}
