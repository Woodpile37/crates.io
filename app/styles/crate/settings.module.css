.email-form {
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap;
    gap: var(--space-s);
    padding: var(--space-s) var(--space-m);
    background-color: white;
    border-radius: var(--space-3xs);
    box-shadow: 0 1px 3px hsla(51, 90%, 42%, .35);
}

.email-input-label {
    font-weight: bold;
}

.email-input {
    width: 400px;
}

.submit-button {
    composes: small yellow-button from '../shared/buttons.module.css';
}

.list {
    background-color: white;
    border-radius: var(--space-3xs);
    box-shadow: 0 1px 3px hsla(51, 90%, 42%, .35);

    > * {
        padding: var(--space-s) var(--space-m);
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-wrap: wrap;
    }

    > * + * {
        border-top: 1px solid hsla(51, 90%, 42%, .25);
    }
}

.email-column {
    width: 25%;
    color: var(--main-color-light);
}

.remove-button {
    composes: small yellow-button from '../shared/buttons.module.css';
}
