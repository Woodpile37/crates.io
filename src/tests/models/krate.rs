use crate::builders::{CrateBuilder, VersionBuilder};
use crate::util::insta::assert_json_snapshot;
use crate::TestApp;
use chrono::{Days, Utc};

#[test]
fn index_metadata() {
    let (app, _, user) = TestApp::init().with_user();
    let user = user.as_model();

    app.db(|conn| {
        let created_at_1 = Utc::now()
            .checked_sub_days(Days::new(14))
            .unwrap()
            .naive_utc();

        let created_at_2 = Utc::now()
            .checked_sub_days(Days::new(7))
            .unwrap()
            .naive_utc();

        let fooo = CrateBuilder::new("foo", user.id)
            .version(VersionBuilder::new("0.1.0"))
            .expect_build(conn);

        let metadata = fooo.index_metadata(conn).unwrap();
        assert_json_snapshot!(metadata);

        let bar = CrateBuilder::new("bar", user.id)
            .version(
                VersionBuilder::new("1.0.0-beta.1")
                    .created_at(created_at_1)
                    .yanked(true),
            )
            .version(VersionBuilder::new("1.0.0").created_at(created_at_1))
            .version(
                VersionBuilder::new("2.0.0")
                    .created_at(created_at_2)
                    .dependency(&fooo, None),
            )
            .version(VersionBuilder::new("1.0.1").checksum("0123456789abcdef"))
            .expect_build(conn);

        let metadata = bar.index_metadata(conn).unwrap();
        assert_json_snapshot!(metadata);
    });
}
