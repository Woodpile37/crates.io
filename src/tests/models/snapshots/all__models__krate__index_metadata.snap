---
source: src/tests/models/krate.rs
expression: metadata
---
[
  {
    "name": "foo",
    "vers": "0.1.0",
    "deps": [],
    "cksum": "                                                                ",
    "features": {},
    "yanked": false
  }
]
